### DISCUSSION TITLE
Read & Write L Vars

### DISCUSSION THREAD

**Discussion by ksi9302** [EvenAR/node-simconnect#106] [Q&A] [answered]
Hey folks!

Following my recent development of PMDG 777, 737 CDUs, I'm trying to make one for my beloved ATR and unfortunately, it doesn't have nice SDKs like the ones PMDG provides..
I believe I'll have to read and write lvars to control the CDU.
Do you think this is possible with simconnect library alone or with WASM involved?

Any ideas are welcome!

**Comment by EvenAR** [comment, marked as answer]
Hi! For sending keypresses, you might want to use input events (aka H-events if I'm not mistaken). [This sample](https://github.com/EvenAR/node-simconnect/blob/ac457842e15fc18f888dc9fea71b77d88893c394/samples/typescript/inputEvents.ts#L47) shows how to retrieve a list of all the available ones. Then you would use [`setInputEvent()`](https://docs.flightsimulator.com/html/Programming_Tools/SimConnect/API_Reference/InputEvents/SimConnect_SetInputEvent.htm) to trigger them.

Annunciators/lights seem to be available as LVARs. I think you can read those in the same way as simvars now. See [SimConnect_AddToDataDefinition](https://docs.flightsimulator.com/html/Programming_Tools/SimConnect/API_Reference/Events_And_Data/SimConnect_AddToDataDefinition.htm). You can find the available LVARs in the MSFS developer menu under Tools -> Behaviors -> LocalVariables. Those starting with `MSATR_MCDU1_****` are probably what you are looking for.

I'm not sure how you can extract the CDU screen text though.. Looks like LVARs are only for numeric values.

**Reply by ksi9302** [reply]
Hi Evan, 

Much appreciate your insight!
Thanks to your reply, I could successfully read all LVARs and input events.
However, I'm still struggling to trigger an input event on the CDU..

I've tried:

1. setInputEvent
```typescript
const res = recv.handle.setInputEvent(1.34373e19, 0); //hash id retrieved following the sample
console.log(res); // 3
```

2. setDataonSimObject
```typescript
 const enum DefinitionID {
    CDU1_CLR,
  }

  recv.handle.addToDataDefinition(
    DefinitionID.CDU1_CLR,
    "L:MSATR_MCDU1_CLR",
    null,
    SimConnectDataType.FLOAT64
  );

  const dataToSet = new RawBuffer(0);
  dataToSet.clear();
  dataToSet.writeFloat64(0);

  recv.handle.setDataOnSimObject(
    DefinitionID.CDU1_CLR,
    SimConnectConstants.OBJECT_ID_USER,
    {
      buffer: dataToSet,
      arrayCount: 0,
      tagged: false,
    }
  );
```

Unfortunately, both method didn't work.
Could you let me know if you see anything suspicious/wrong in the code?

Thanks!

p.s. with extracting the CDU screen, I believe the only method available as of now is capturing the pop up window and sending the image over websocket which I kind of succeded..

**Reply by EvenAR** [reply]
The hash IDs returned by SimConnect are a bit tricky since they are of type `unsigned __int64` (according to the original SimConnect.h file for C) which the standard node.js `number` type cannot accurately represent. This is why node-simconnect uses the `Long` type for the hash IDs. Looking at the ID in you sample code I assume that you are using `Long.toNumber()` or `Long.toInt()` which will remove some digits from the ID. One option is to store the IDs you get from node-simconnect without converting them. 

I realise that the [`BigInt`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt) type also could work. I guess you could convert the values to `BigInt` which could be handier, but you would still need to convert them back to `Long` when sending the event.. Perhaps I should update `node-simconnect` to use `BigInt` instead of the not so standard `Long` type 🤔 

> p.s. with extracting the CDU screen, I believe the only method available as of now is capturing the pop up window and sending the image over websocket which I kind of succeded..

Cool! 🙌

**Reply by ksi9302** [reply]
Thanks Evan, ahh, I was so naive.
I could get the setInputEvent working by doing the below.

```typescript
  handle.on("inputEventsList", recvEnumerateInputEvents => {
    recvEnumerateInputEvents.inputEventDescriptors.forEach(e => {
      if (e.name == "INSTRUMENT_MCDU1_Key_CLR") {
        handle.setInputEvent(e.inputEventIdHash, 1);
        handle.setInputEvent(e.inputEventIdHash, 0);
        console.log(e.inputEventIdHash);
      }
    });
  });
```

However, after identifying the actual long value which was:

Long { low: -625217470, high: -1166357638, unsigned: true }

I tried constructing a variable with the exact same value using the long library.

```typescript
  const clr_button = new Long(-625217470, -1166357638, true); // tried Long.fromXX methods as well..
  handle.setInputEvent(clr_button, 1);
  handle.setInputEvent(clr_button, 0);
```

Which sadly didn't work due to illegal value type error (not an integer or Long)
For now, I'll just store the original returned ids I need in a variable, but it would be nice to have a method to convert the value to an acceptable long type within the library..! (Or maybe there already is..?)

Thanks!

**Reply by EvenAR** [reply]
Yes, I also faced that error when I did some testing now🤔 

I just published `node-simconnect@4.0.0-alpha.0` which uses bigint so you don't need to deal with the Long library. Hopefully that will make things easier. bigints can easily be converted to strings and then back using for instance `BigInt("12259035040569282066")`. You can even use this syntax if you are using ECMAScript 2020 or newer: 
```ts
handle.setInputEvent(12259035040569282066n, 1) // notice the trailing "n"
```
Feel free to give it a try!

**Reply by ksi9302** [reply]
SUPER,

It now works perfectly and it's almost too easy now to send any input.

To sum up the solution,

1. Reading LVARs
```typescript
  handle.addToDataDefinition(
    AIRCRAFT_DATA_DEFINITION_ID,
    "L:MSATR_MCDU1_MSG_LT",
    null,
    SimConnectDataType.FLOAT64,
    0,
    SimConnectConstants.UNUSED
  );

  handle.requestDataOnSimObject(
    AIRCRAFT_DATA_REQUEST_ID,
    AIRCRAFT_DATA_DEFINITION_ID,
    SimConnectConstants.OBJECT_ID_USER,
    SimConnectPeriod.ONCE
  );
```

2. Writing input events
```typescript
  const INSTRUMENT_MCDU1_Key_CLR = "13437276166729494594";

  // sends short click event, only sending an event with 1 will result in long click.
  handle.setInputEvent(BigInt(INSTRUMENT_MCDU1_Key_CLR), 1);
  handle.setInputEvent(BigInt(INSTRUMENT_MCDU1_Key_CLR), 0);
```

Much much thanks Evan!
And, Here I come ATR!

**Reply by EvenAR** [reply]
Great! I haven’t found any documentation on  what the hash IDs are based on btw. I suspect that they are a result of some hashed HTML code, and that they could possibly change when the addon is updated at some point. Instead of hard coding the IDs it might be a better idea to fetch all every time your app connects and look up the IDs by the names..

**Reply by ksi9302** [reply]
Dang...! I didn't think it would change, more work..!! Thanks for the heads up mate!
